### ISSUE TITLE
'torch.masked_select' behaves differently depending on the device of inputs.

### ISSUE THREAD

**Issue by jup-proxem** [pytorch/pytorch#41473]
## 🐛 Bug

Outputs from torch.masked_select are different when input tensors' device is cpu and when it's cuda.

## To Reproduce

1. candidate = torch.tensor([[[-1, -2], [0, 1], [2, 3], [4, 5], [6, 7], [8, 9]],
                              [[-11, -12], [10, 11], [12, 13], [14, 15], [16, 17], [18, 19]],
                              [[-101, -102], [100, 101], [102, 103], [104, 105], [106, 107], [108, 109]]])
1. mask = torch.tensor([[0, 1, 1, 1, 1, 0],
                              [0, 1, 1, 0, 1, 1],
                              [0, 1, 1, 0, 1, 1]], dtype=torch.bool)
1. result_cpu = torch.masked_select(candidate.permute(2, 0, 1), mask)
1. result_gpu = torch.masked_select(candidate.to('cuda').permute(2, 0, 1), mask.to('cuda'))
1. assert not (result_cpu != result_gpu.cpu()).any()

## Expected behavior

Outputs (result_cpu and result_gpu) should be identical.
Gettting : 
result_cpu = tensor([  0, 1, 2, 3, 4, 5, 6, 7, 10, 11, 12, 13, 16, 17, 18, 19, 100, 101, 102, 103, 106, 107, 108, 109])
result_gpu = tensor([ 0, 2, 4, 6, 10, 12, 16, 18, 100, 102, 106, 108, 1, 3, 5, 7, 11, 13, 17, 19, 101, 103, 107, 109], device='cuda:0')

## Environment
PyTorch version: 1.7.0a0+c71ec1c
Is debug build: No
CUDA used to build PyTorch: 10.1

OS: Ubuntu 16.04.6 LTS
GCC version: (Ubuntu 5.4.0-6ubuntu1~16.04.12) 5.4.0 20160609
CMake version: version 3.14.0

Python version: 3.8
Is CUDA available: Yes
CUDA runtime version: 10.1.105
GPU models and configuration: GPU 0: Tesla K40m
Nvidia driver version: 440.33.01
cuDNN version: /usr/lib/x86_64-linux-gnu/libcudnn.so.7.1.3

Versions of relevant libraries:
[pip3] numpy==1.18.5
[pip3] torch==1.7.0a0+c71ec1c
[conda] blas                      1.0                         mkl    defaults
[conda] magma-cuda101             2.5.2                         1    pytorch
[conda] mkl                       2020.1                      217    defaults
[conda] mkl-include               2020.1                      217    defaults
[conda] mkl-service               2.3.0            py38he904b0f_0    defaults
[conda] mkl_fft                   1.1.0            py38h23d657b_0    defaults
[conda] mkl_random                1.1.1            py38h0573a6f_0    defaults
[conda] numpy                     1.18.5           py38ha1c710e_0    defaults
[conda] numpy-base                1.18.5           py38hde5b4d6_0    defaults
[conda] torch                     1.7.0a0+c71ec1c          pypi_0    pypi

cc @ezyang @gchanan @zou3519

**Comment by ezyang** [comment]
silent correctness bug, we should fix this

**Comment by vinnamkim** [comment]
@jup-proxem

In the reproduction stpes, `result_gpu` is selected by `mask_test`. I guess it should be selected by `mask` same as `result_cpu`. It would be there is a mistake like `mask != mask_test`. The following codes have no problem.

```
import torch

candidate = torch.tensor([[[-1, -2], [0, 1], [2, 3], [4, 5], [6, 7], [8, 9]],
[[-11, -12], [10, 11], [12, 13], [14, 15], [16, 17], [18, 19]],
[[-101, -102], [100, 101], [102, 103], [104, 105], [106, 107], [108, 109]]])
mask = torch.tensor([[0, 1, 1, 1, 1, 0],
[0, 1, 1, 0, 1, 1],
[0, 1, 1, 0, 1, 1]], dtype=torch.bool)
result_cpu = torch.masked_select(candidate.permute(2, 0, 1), mask)
result_gpu = torch.masked_select(candidate.to('cuda').permute(2, 0, 1), mask.to('cuda'))

assert not (result_cpu != result_gpu.cpu()).any()
```

**Comment by jup-proxem** [comment]
> In the reproduction stpes, `result_gpu` is selected by `mask_test`. I guess it should be selected by `mask` same as `result_cpu`. It would be there is a mistake like `mask != mask_test`. The following codes have no problem.

 @vinnamkim you're right, it was a typo on my end, I fixed it.

**Comment by jup-proxem** [comment]
@ezyang 

If you closed the issue because you fixed the bug, thank you and ignore the rest of my message.

Otherwise,
The typo in the reproduction steps was not responsible for the bug, by "I fixed it" I meant in the bug report. I still have an assertion error when following the reproduction steps.

**Comment by zou3519** [comment]
I ran the above example on PyTorch 1.5.1 and again on the 1.6 release branch. It looks like on PyTorch 1.5.1, there is no problem -- the cpu and gpu versions of masked_select produce the same answer. However, on the release branch, the issue is present.

**Comment by zou3519** [comment]
Did some more investigation. It looks like masked_select on CPU is giving bogus results on non-contiguous tensors
```
x = torch.randn(3, 3)
mask = torch.randint(0, 2, (3, 3), dtype=torch.bool)
cpu_output = torch.masked_select(x.t(), mask)
cuda_output = torch.masked_select(x.cuda().t(), mask)
assert torch.allclose(cpu_output, cuda_output.cpu())
```

**Comment by zou3519** [comment]
Bisected to https://github.com/pytorch/pytorch/issues/33269

**Comment by zou3519** [comment]
It looks like the bug only gets triggered if the input tensor has < at::internal::GRAIN_SIZE elements (32768). I verified this experimentally. The codepath at https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/TensorAdvancedIndexing.cpp#L711-L724 seems to be where the incorrect behavior is, but skimming through the code I am not sure what is up

**Comment by gchanan** [comment]
the `out=` non-contiuous handling also looks wrong, whether or not the input tensor is smaller than the grain size:

small:
```
>>> x=torch.randn(30,30)
>>> mask = torch.randn(30, 30) < 0
>>> torch.masked_select(x, mask).shape
torch.Size([477])
>>> o=torch.zeros(477*2)[::2]
>>> a=torch.masked_select(x, mask)
torch.masked_select(x, mask, out=o)
>>> torch.allclose(a, o)
False
```

big:
```
>>> y = torch.randn(200, 200)
>>> bmask = torch.randn(200, 200) < 0
>>> a=torch.masked_select(y, bmask)
>>> a.shape
torch.Size([19926])
>>> o2 = torch.zeros(19926*2)[::2]
>>> torch.masked_select(y, bmask, out=o2)
>>> torch.allclose(a, o2)
False
```

**Comment by gchanan** [comment]
Here's a PR that does the revert and fixes up an issue with THByteTensor_sumall not existing anymore: https://github.com/pytorch/pytorch/pull/41828

**Comment by zou3519** [comment]
The particular issue is closed by #41828. However we would like to "un-revert the revert and fix the problem", that work is happening in #41841.
